File: src/Solutions/Euler623.hs
{-# LANGUAGE Strict #-}
module Solutions.Euler623 where
import Utils.Memo

pplus x y = (x + y) `mod` (10^9 + 7)
psum = foldr pplus 0

-- Counts lambda terms with exact length len and vars unbounded variables
exactLambdaCount :: (Int, Int) -> StrictMemo (Int, Int) Int Int
exactLambdaCount (len, vars)
    | len <= 0 = return 0
    | len == 1 = return vars
    | otherwise = do
        application <- fmap psum $ forM [1..len - 3] $ \m ->
            (*) <$> memo exactLambdaCount (m, vars)
                <*> memo exactLambdaCount (len - 2 - m, vars)
        abstraction <- memo exactLambdaCount (len - 5, vars + 1)
        return $ pplus application abstraction

lambdaCount :: Int -> StrictMemo (Int, Int) Int Int
lambdaCount n = psum <$> forM [6..n] (\l -> memo exactLambdaCount (l, 0))

runLambdaCount :: Int -> Int
runLambdaCount = startEvalMemo . lambdaCount

euler623 :: IO String
euler623 = return $ show $ runLambdaCount 2000

File: rust/src/solutions/euler623.rs
use cached::proc_macro::cached;

fn psum(a: i64, b: i64) -> i64 {
    (a + b) % 1000000007
}

#[cached]
fn exact_lambda_count(len: i64, vars: i64) -> i64 {
    if len <= 0 {
        0
    } else if len == 1 {
        vars
    } else {
        let application = (1..len - 2)
            .map(|m| exact_lambda_count(m, vars) * exact_lambda_count(len - 2 - m, vars))
            .fold(0, psum);
        let abstraction = exact_lambda_count(len - 5, vars + 1);
        psum(application, abstraction)
    }
}

fn lambda_count(max_len: i64) -> i64 {
    (6..=max_len)
        .map(|len| exact_lambda_count(len, 0))
        .fold(0, psum)
}

pub fn euler623() -> String {
    lambda_count(2000).to_string()
}
